pkg/kubelet/kubeletconfig/util/codec/codec.go

```diff
@@ -72,16 +77,50 @@ func NewYAMLEncoder(groupName string) (runtime.Encoder, error) {
 		return nil, fmt.Errorf("no enabled versions for group %q", groupName)
 	}
 
-	// the "best" version supposedly comes first in the list returned from legacyscheme.Registry.EnabledVersionsForGroup
+	// the "best" version supposedly comes first in the list returned from legacyscheme.Registry.EnabledVersionsForGroup.
 	return legacyscheme.Codecs.EncoderForVersion(info.Serializer, versions[0]), nil
 }
 
-// DecodeKubeletConfiguration decodes a serialized KubeletConfiguration to the internal type
+// newLenientSchemeAndCodecs returns a scheme that has only v1beta1 registered into
+// it and a CodecFactory with strict decoding disabled.
+func newLenientSchemeAndCodecs() (*runtime.Scheme, *serializer.CodecFactory, error) {
+	lenientScheme := runtime.NewScheme()
+	if err := kubeletconfig.AddToScheme(lenientScheme); err != nil {
+		return nil, nil, fmt.Errorf("failed to add kubelet config API to lenient scheme: %v", err)
+	}
+	if err := kubeletconfigv1beta1.AddToScheme(lenientScheme); err != nil {
+		return nil, nil, fmt.Errorf("failed to add kubelet config v1beta1 API to lenient scheme: %v", err)
+	}
+	lenientCodecs := serializer.NewCodecFactory(lenientScheme, serializer.DisableStrict)
+	return lenientScheme, &lenientCodecs, nil
+}
+
+// DecodeKubeletConfiguration decodes a serialized KubeletConfiguration to the internal type.
 func DecodeKubeletConfiguration(kubeletCodecs *serializer.CodecFactory, data []byte) (*kubeletconfig.KubeletConfiguration, error) {
-	// the UniversalDecoder runs defaulting and returns the internal type by default
+	var (
+		obj runtime.Object
+		gvk *schema.GroupVersionKind
+	)
+
+	// The UniversalDecoder runs defaulting and returns the internal type by default.
 	obj, gvk, err := kubeletCodecs.UniversalDecoder().Decode(data, nil, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to decode, error: %v", err)
+		// Try strict decoding first. If that fails decode with a lenient
+		// decoder, which has only v1beta1 registered, and log a warning.
+		// The lenient path is to be dropped when support for v1beta1 is dropped.
+		if !runtime.IsStrictDecodingError(err) {
+			return nil, errors.Wrap(err, "failed to decode")
+		}
+		klog.Warningf("using lenient decoding as strict decoding failed: %v", err)
```

Comment:
it seems like we should only print the warning if we successfully decode with the lenient codec and continue with that object (otherwise, a v1 config with duplicate keys will always print this warning)

also, it seems like we should return the original strict error when lenient decoding also fails (otherwise, a v1 config would return an unknown kind error because we are not registering v1 into the lenient codec)

```go
var lenientErr error
_, lenientCodecs, lenientErr := newLenientSchemeAndCodecs()
if lenientErr != nil {
  return nil, lenientErr
}

obj, gvk, lenientErr = lenientCodecs.UniversalDecoder().Decode(data, nil, nil)
if lenientErr != nil {
  // lenient decoding also failed, return the original strict error
  return nil, fmt.Errorf("failed to decode, error: %v", err)
}
// continue with the object decoded leniently, but warn
klog.Warningf("using lenient decoding as strict decoding failed: %v", err)
```

Reply:
If we did the latter (return the original strict error), could that mask the real problem in the v1?
For instance, earlier in the file I have an unrecognized key, and later in the file I have a syntax error.

Reply:
Either way it's a confusing error...

Reply:
> If we did the latter (return the original strict error), could that mask the real problem in the v1?

v1 should always be strict, and should always fail the lenient decoder (because it only has v1beta1 registered), right?

Reply:
@liggitt fixed